pkg/resources/cloudcontroller/vsphere.go

```diff
@@ -0,0 +1,162 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	vsphereCPIControllerManager = "vsphere-cloud-controller-manager"
+)
+
+var (
+	vsphereCPIResourceRequirements = corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceMemory: resource.MustParse("100Mi"),
+			corev1.ResourceCPU:    resource.MustParse("200m"),
+		},
+		Limits: corev1.ResourceList{
+			corev1.ResourceMemory: resource.MustParse("512Mi"),
+			corev1.ResourceCPU:    resource.MustParse("500m"),
+		},
+	}
+)
+
+func vsphereDeploymentCreator(data *resources.TemplateData) reconciling.NamedDeploymentCreatorGetter {
+	return func() (string, reconciling.DeploymentCreator) {
+		return vsphereCPIControllerManager, func(dep *appsv1.Deployment) (*appsv1.Deployment, error) {
+			dep.Name = vsphereCPIControllerManager
+			dep.Labels = resources.BaseAppLabels(vsphereCPIControllerManager, nil)
+
+			dep.Spec.Selector = &metav1.LabelSelector{
+				MatchLabels: resources.BaseAppLabels(vsphereCPIControllerManager, nil),
+			}
+
+			podLabels, err := data.GetPodTemplateLabels(vsphereCPIControllerManager, dep.Spec.Template.Spec.Volumes, map[string]string{
+				"component": "cloud-controller-manager",
+				"tier":      "control-plane",
+			})
+			if err != nil {
+				return nil, err
+			}
+			dep.Spec.Template.ObjectMeta = metav1.ObjectMeta{
+				Labels: podLabels,
+			}
+
+			dep.Spec.Template.Spec.DNSPolicy, dep.Spec.Template.Spec.DNSConfig, err =
+				resources.UserClusterDNSPolicyAndConfig(data)
+			if err != nil {
+				return nil, err
+			}
+			dep.Spec.Template.Spec.AutomountServiceAccountToken = pointer.BoolPtr(false)
+			dep.Spec.Template.Spec.HostNetwork = true
+
+			version, err := getVsphereCPIVersion(data.Cluster().Spec.Version)
+			if err != nil {
+				return nil, err
+			}
+			openvpnSidecar, err := vpnsidecar.OpenVPNSidecarContainer(data, openvpnClientContainerName)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get openvpn sidecar: %v", err)
+			}
+			containers, err := getCPIContainerList(version, data)
+			if err != nil {
+				return nil, err
+			}
+			dep.Spec.Template.Spec.Containers = append(containers, *openvpnSidecar)
+
+			dep.Spec.Template.Spec.Volumes = append(getVolumes(), corev1.Volume{
+				Name: resources.CloudConfigConfigMapName,
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: resources.CloudConfigConfigMapName,
+						},
+					},
+				},
+			})
+
+			return dep, nil
+		}
+	}
+}
+
+func getCPIContainerList(version string, data *resources.TemplateData) ([]corev1.Container, error) {
+	controllerManagerImage := fmt.Sprintf("%s/cloud-provider-vsphere/cpi/release/manager:v%s", data.ImageRegistry(resources.RegistryGCR), version)
+
+	containerList := []corev1.Container{
+		{
+			Name:  vsphereCPIControllerManager,
+			Image: controllerManagerImage,
+			Args: []string{
```

Comment:
We should also set the `--cluster-name` flag to avoid problems similar to #7229.

We can do it in two ways:

1) Set it unconditionally
2) Set it in the same way as for OpenStack (by respecting the `ccmClusterName` cluster feature): https://github.com/kubermatic/kubermatic/blob/43d0e7bbaa8c445ca9dd5931a3466e40ebf199a4/pkg/resources/cloudcontroller/openstack.go#L138-L140

I think we should go with the second approach. Currently, the vSphere CCM will be deployed only for the new user clusters, so the first approach should work. However, we'll also migrate existing clusters later, which don't have the cluster name set, so the second approach can help us avoid potential problems that we encountered on OpenStack.

Note that if you go with the second approach, you also need to update this function, so the cluster feature is enabled by default for new user clusters: https://github.com/kubermatic/kubermatic/blob/43d0e7bbaa8c445ca9dd5931a3466e40ebf199a4/pkg/resources/cloudcontroller/util.go#L55-L64